lib/dynamodb/dynamodb_request_queue.c

```diff
@@ -243,6 +221,63 @@ extractcapacity(struct http_response * res, double * pcap)
 	return (-1);
 }
 
+/* Extract a DynamoDB error string from an HTTP 400 response. */
+static int
+extracterror(struct http_response * res, char ** errstr)
+{
+	const uint8_t * buf = res->body;
+	const uint8_t * end = res->body + res->bodylen;
+	size_t len;
+
+	/* Look for __type. */
+	buf = json_find(buf, end, "__type");
+
+	/*
+	 * We should be pointing at the opening double-quote of a string;
+	 * move past it.
+	 */
+	if (buf[0] != '"')
+		goto bad;
+	buf++;
+
+	/* Find a '#' inside the string and move past it. */
+	while (buf[0] != '#') {
```

Comment:
It might be worth adding a comment that `json_find()` is guaranteed to return a valid string which is enlosed by `"..."`.  Otherwise this `while` statement can end up reading unallocated memory.

(For that matter, I'm surprised that clang-scan doesn't complain about this, but there's no complaints in clang 11, at least.)

Reply:
Oops.  Actually no, `json_find` isn't guaranteed to return a string enclosed by anything; it just looks for the start of an item.  If the buffer is truncated we'll absolutely run past the end of it.

Fix added.